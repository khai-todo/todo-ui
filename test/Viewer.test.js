import renderer from 'react-test-renderer'
import Subject from '../src/Viewer'
import snap from './lib/snapshot'
import data from './lib/data'

it('empty rendering', () => snap(<Subject />))

describe('when passing data', () => {
  const {correct, ...incorrect} = data

  it('accept correct data', () => snap(<Subject data={correct} />))

  describe('that is incorrect', () => Object.entries(incorrect).forEach(([key, val]) => {
    describe(`that came from ${key}.yaml`, () => {
      it('should throw an error when no alternate construct is provided', () => {
        expect(() => <Subject data={val} />).toThrow()
      })

      it('should render alternate construct when it is provided', () => {
        const OnError = () => <output>
          <p>An Error occurred.</p>
          <p><code><pre>{
            '\n' + JSON.stringify({key, val}, undefined, 2) + '\n'
          }</pre></code></p>
        </output>

        expect(
          renderer.create(
            <Subject data={val} OnError={OnError} />
          ).toJSON()
        ).toBe(
          renderer.create(
            <OnError />
          ).toJSON()
        )
      })
    })
  }))
})
