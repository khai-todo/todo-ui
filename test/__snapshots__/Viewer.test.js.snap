// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when passing data accept valid data Nested x1 1`] = `
<ul>
  <li>
    <details>
      <summary>
        <span>
          Top
        </span>
      </summary>
    </details>
  </li>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Bottom 1
          </span>
        </summary>
      </details>
    </li>
  </ul>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Bottom 2
          </span>
        </summary>
      </details>
    </li>
  </ul>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Bottom 3
          </span>
        </summary>
      </details>
    </li>
  </ul>
</ul>
`;

exports[`when passing data accept valid data Nested x2 1`] = `
<ul>
  <li>
    <details>
      <summary>
        <span>
          Top
        </span>
      </summary>
    </details>
  </li>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Middle 1 — Simple
          </span>
        </summary>
      </details>
    </li>
  </ul>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Middle 2 — Has content
          </span>
        </summary>
        This is content of Middle 2
      </details>
    </li>
  </ul>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Middle 3 — Has children
          </span>
        </summary>
      </details>
    </li>
    <ul>
      <li>
        <details>
          <summary>
            <span>
              Bottom 3.1
            </span>
          </summary>
        </details>
      </li>
    </ul>
    <ul>
      <li>
        <details>
          <summary>
            <span>
              Bottom 3.2
            </span>
          </summary>
        </details>
      </li>
    </ul>
    <ul>
      <li>
        <details>
          <summary>
            <span>
              Bottom 3.3
            </span>
          </summary>
        </details>
      </li>
    </ul>
  </ul>
  <ul>
    <li>
      <details>
        <summary>
          <span>
            Middle 4 — Has content and children
          </span>
        </summary>
        This is content of Middle 4
      </details>
    </li>
    <ul>
      <li>
        <details>
          <summary>
            <span>
              Bottom 4.1
            </span>
          </summary>
        </details>
      </li>
    </ul>
    <ul>
      <li>
        <details>
          <summary>
            <span>
              Bottom 4.2
            </span>
          </summary>
        </details>
      </li>
    </ul>
    <ul>
      <li>
        <details>
          <summary>
            <span>
              Bottom 4.3
            </span>
          </summary>
        </details>
      </li>
    </ul>
  </ul>
</ul>
`;

exports[`when passing data accept valid data Without children 1`] = `
<ul>
  <li>
    <details>
      <summary>
        <span>
          Title
        </span>
      </summary>
      Content
    </details>
  </li>
</ul>
`;

exports[`when passing data accept valid data Without children and content 1`] = `
<ul>
  <li>
    <details>
      <summary>
        <span>
          Title
        </span>
      </summary>
    </details>
  </li>
</ul>
`;

exports[`when passing data handle invalid data without error handler throws correct error message 1`] = `
Object {
  "error": [TypeError: Invalid type of data (data is null or undefined)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 2`] = `
Object {
  "error": [TypeError: Invalid type of data (typeof data !== 'object')],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 3`] = `
Object {
  "error": [TypeError: Invalid type of data (typeof data !== 'object')],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 4`] = `
Object {
  "error": [TypeError: Invalid type of data (data is an array)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 5`] = `
Object {
  "error": [Error: Missing Field (title)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 6`] = `
Object {
  "error": [Error: Invalid field (title: Provided value is not a string)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 7`] = `
Object {
  "error": [Error: Invalid field (title: Provided value is not a string)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 8`] = `
Object {
  "error": [Error: Invalid field (children: Provided value is neither falsy nor an array)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 9`] = `
Object {
  "error": [Error: Invalid field (children: Provided value is neither falsy nor an array)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 10`] = `
Object {
  "error": [Error: Invalid field (children: Provided value is neither falsy nor an array)],
}
`;

exports[`when passing data handle invalid data without error handler throws correct error message 11`] = `
Object {
  "error": [TypeError: Invalid type of data (typeof data !== 'object')],
}
`;
